### PR TITLE
[TrimmableTypeMap] SafeJavaCollectionFactory per-container construction

### PR DESCRIPTION
**Pull request by simonrozsival**
> [!NOTE]
> Replacement for #12131 because its fork head repository was deleted. This branch now lives in `dotnet/android`.

## What

Reworks the trimmable-typemap Java-collection converter (`SafeJavaCollectionFactory`) so the AOT-safe construction of `JavaList<T>` / `JavaCollection<T>` / `JavaDictionary<K,V>` is explicit and easy to reason about, and hardens a couple of edge cases. Net change touches the two implementation files, `SafeJavaCollectionFactory.cs` and `ValueTypeFactory.cs`, plus a targeted `JavaConvertTest` regression test (the factory name, public entry point name, and `JavaConvert.cs` are unchanged vs `main`).

### Construction split into explicit, non-overlapping paths

Container kind is resolved once when the converter is selected, producing a container-specific closure (`CreateListFromJniHandle` / `CreateCollectionFromJniHandle` / `CreateDictionaryFromJniHandle`) that captures only its exact generic arguments. Each method handles exactly one wrapper:

- **reference element/key/value args** → direct `new` for the concrete `IJavaPeerable` rooting anchors; other reference types use `MakeGenericType(...)` + `Activator.CreateInstance`, riding NativeAOT's `__Canon` canonical templates;
- **mapped primitive/nullable args** → `ValueTypeFactory`, rooting the exact instantiation with a direct `new` (no reflection);
- **any other value type** (e.g. a custom struct) → this factory declines the conversion so `JavaConvert` preserves its existing untyped collection fallback, covered for list, collection, and dictionary interface targets by a trimmable-path regression test.

### Concrete-literal rooting (replaces `[DynamicallyAccessedMembers]` for reference wrappers)

Each reference path has an `if (elementType == typeof(IJavaPeerable))` branch that constructs the concrete `JavaList<IJavaPeerable>` / `JavaCollection<IJavaPeerable>` / `JavaDictionary<IJavaPeerable,IJavaPeerable>` literal. That branch is effectively unreachable at runtime, but being a statically reachable reference to the concrete instantiation it forces the trimmer/ILC to keep the `(IntPtr, JniHandleOwnership)` activation constructor of the shared `__Canon` template. The `else` branch reuses exactly that canonical constructor for every other reference-argument instantiation. This roots the constructor through real code rather than relying on `[DAM]` flowing through reflection.

`ValueTypeFactory`'s mixed reference/value dictionaries (`JavaDictionary<string,int>`, …) use the analogous `CreateReferenceMixedDictionary<JavaDictionary<IJavaPeerable,T>>` / `<JavaDictionary<T,IJavaPeerable>>` exemplars to root the `JavaDictionary<__Canon,T>` / `JavaDictionary<T,__Canon>` templates, replacing the previous static rooting-token fields. Value/value dictionaries continue to be rooted by the statically-emitted GVM cross-product (`new JavaDictionary<TKey,T>()`).

### Hardening

- **Reject open / partially-open constructed targets** (`IList<>`, `IDictionary<T,int>`, …) via a `ContainsGenericParameters` check up front, so they fail cleanly instead of throwing `ArgumentException` during activation.

### Suppressions

The reflective paths carry the minimum verified suppressions: `IL3050` (RequiresDynamicCode) and `IL2071` on `MakeGenericType`, plus `IL2072` on the `Activator.CreateInstance` calls whose constructed types ride the rooted canonical templates. The `IL2071` justification states the real cause — `MakeGenericType` cannot prove the individual runtime type arguments satisfy the `[DAM(Constructors)]` requirements on the wrapper's element parameters, requirements only exercised by the dynamic-code path (the trimmable path never activates elements through the wrapper).

## Validation

- ✅ Builds locally (`Mono.Android`, Debug, `net11.0`) with no new warnings and no IL trim/AOT analysis errors.
- ✅ Added NativeAOT reference-only round trips for list, collection, and dictionary targets; expected wrapper types remain open generic definitions so the closed wrappers must be activated through the rooted canonical templates.

Co-authored-by: Copilot <223556219+Copilot@users.noreply.github.com>

### CODE REVIEW

File: src/Mono.Android/Java.Interop/SafeJavaCollectionFactory.cs [hunk 1]
```diff
@@ -61,170 +92,149 @@ static bool IsSupportedCollectionShape (CollectionShape shape)
...
-			return true;
+		if (elementType == typeof (IJavaPeerable)) {
+			// Concrete-literal rooting branch. Taken only when marshaling the JavaList<IJavaPeerable> shape itself
+			// (uncommon), its main purpose is to keep the reflection metadata + invoke stub of the (IntPtr,
+			// JniHandleOwnership) constructor on the JavaList<__Canon> template alive for the trimmer/ILC. The else
+			// branch reuses that same canonical constructor for every other JavaList<referenceType>.
+			result = Activator.CreateInstance (typeof (JavaList<IJavaPeerable>), ActivationConstructorBinding, binder: null, args: [handle, transfer], culture: CultureInfo.InvariantCulture);
+		} else {
+			var listType = typeof (JavaList<>).MakeGenericType (elementType);
+			result = Activator.CreateInstance (listType, ActivationConstructorBinding, binder: null, args: [handle, transfer], culture: CultureInfo.InvariantCulture);
```
[the remark this answers is not in the document]
Comment (jonathanpeppers) [reply src/Mono.Android/Java.Interop/SafeJavaCollectionFactory.cs]: Addressed in d41a8ae3c by adding fail-fast null guards to all three reflective construction paths and marking successful TryCreate results as non-null.

File: src/Mono.Android/Java.Interop/SafeJavaCollectionFactory.cs [hunk 2]
```diff
@@ -61,170 +92,158 @@ static bool IsSupportedCollectionShape (CollectionShape shape)
...
 		if (handle == IntPtr.Zero) {
 			return null;
 		}
 
-		if (TryCreateFromMappedValueTypeFactories (shape, handle, transfer, out var collection)) {
-			return collection;
+		object? result;
+		if (TryCreateListFromJniHandle (genericDefinition, arguments, handle, transfer, out result)
+			|| TryCreateCollectionFromJniHandle (genericDefinition, arguments, handle, transfer, out result)
+			|| TryCreateDictionaryFromJniHandle (genericDefinition, arguments, handle, transfer, out result)) {
```
Comment (rolfbjarne) [inline src/Mono.Android/Java.Interop/SafeJavaCollectionFactory.cs]: `arguments` must have exactly 1 element for List and Collection, and exactly 2 elements for Dictionary. This doesn't seem to be validated anywhere.

This can be used to split these calls (only try to create a List/Collection if `arguments` has 1 element, and only try to create a Dictionary if `arguments` has 2 elements). The signature for the TryCreate methods could also change, and instead of taking "Type[] arguments", take "Type elementType" for List/Collection. and "Type keyType, Type valueType" for Dictionary - which would serve as validation.

File: src/Mono.Android/Java.Interop/SafeJavaCollectionFactory.cs [hunk 3]
```diff
@@ -61,170 +92,158 @@ static bool IsSupportedCollectionShape (CollectionShape shape)
...
 
-		if (genericDefinition == typeof (ICollection<>) || genericDefinition == typeof (JavaCollection<>)) {
-			shape = new CollectionShape (JavaCollectionKind.Collection, targetType.GetGenericArguments ());
-			return true;
+		if (elementType == typeof (IJavaPeerable)) {
+			// Concrete-literal rooting branch. Taken only when marshaling the JavaList<IJavaPeerable> shape itself
+			// (uncommon), its main purpose is to keep the reflection metadata + invoke stub of the (IntPtr,
+			// JniHandleOwnership) constructor on the JavaList<__Canon> template alive for the trimmer/ILC. The else
+			// branch reuses that same canonical constructor for every other JavaList<referenceType>.
+			result = Activator.CreateInstance (typeof (JavaList<IJavaPeerable>), ActivationConstructorBinding, binder: null, args: [handle, transfer], culture: CultureInfo.InvariantCulture);
```
Comment (rolfbjarne) [inline src/Mono.Android/Java.Interop/SafeJavaCollectionFactory.cs]: I admit I'm probably missing something, but it seems this could be replaced with:


```suggestion
			result = new JavaList<IJavaPeerable (handle, transfer);
```

File: src/Mono.Android/Java.Interop/SafeJavaCollectionFactory.cs (hunk 2 above)
Comment (jonathanpeppers) [reply src/Mono.Android/Java.Interop/SafeJavaCollectionFactory.cs]: Addressed in 3ae5a04fe. Container dispatch now happens once when selecting the converter: list/collection closures capture one `Type`, and dictionary closures capture two. This removes the `Type[]` and three chained `TryCreate` calls from the conversion hot path while encoding the expected arity in each method signature.

File: src/Mono.Android/Java.Interop/SafeJavaCollectionFactory.cs (hunk 3 above)
Comment (jonathanpeppers) [reply src/Mono.Android/Java.Interop/SafeJavaCollectionFactory.cs]: Addressed in 3ae5a04fe, including the analogous collection and dictionary branches. Each concrete `IJavaPeerable` anchor now uses direct construction and returns immediately; reflection remains only for runtime reference-type arguments.

**Review by rolfbjarne — APPROVED** [review]

File: src/Mono.Android/Java.Interop/ValueTypeFactory.cs [hunk 4]
```diff
@@ -97,16 +94,62 @@ internal override IDictionary CreateDictionary (ValueTypeFactory valueFactory, I
...
+		JniHandleOwnership transfer)
+	{
+		var exemplarType = typeof (TExemplar);
+		Debug.Assert (exemplarType.IsGenericType && !exemplarType.IsGenericTypeDefinition);
+
+		var definition = exemplarType.GetGenericTypeDefinition ();
+		var dictionaryType = definition.MakeGenericType (arguments);
+		var instance = Activator.CreateInstance (
+			dictionaryType,
+			BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Instance,
```
[the remark this answers is not in the document]
Comment (jonathanpeppers) [reply src/Mono.Android/Java.Interop/ValueTypeFactory.cs]: Addressed in 9e93dfa30. `ActivationConstructorBinding` is now internal and reused by `ValueTypeFactory.CreateReferenceMixedDictionary`.

File: tests/Mono.Android-Tests/Mono.Android-Tests/Java.Interop/JavaConvertTest.cs [hunk 5]
```diff
@@ -221,6 +221,37 @@ public void FromJniHandle_IListByte ()
...
+		[TestCase (typeof (IList<DateTime>), false)]
+		[TestCase (typeof (ICollection<DateTime>), false)]
+		[TestCase (typeof (IDictionary<DateTime, string>), true)]
+		[TestCase (typeof (IDictionary<string, DateTime>), true)]
+		[TestCase (typeof (IList<UnsupportedValueType>), false)]
+		[TestCase (typeof (ICollection<UnsupportedValueType>), false)]
+		[TestCase (typeof (IDictionary<UnsupportedValueType, string>), true)]
+		[TestCase (typeof (IDictionary<string, UnsupportedValueType>), true)]
+		[Category ("NativeAOTTrimmable")]
+		public void FromJniHandle_UnsupportedValueTypeUsesUntypedFallback (Type targetType, bool dictionary)
```
[the remark this answers is not in the document]
Comment (jonathanpeppers) [reply tests/Mono.Android-Tests/Mono.Android-Tests/Java.Interop/JavaConvertTest.cs:254]: Addressed in 9e93dfa30. Added list, collection, and dictionary reference-only round trips through `JavaConvert`; the assertions keep the expected wrappers as open generic definitions so NativeAOT must construct the closed wrappers through the canonical templates. No separate trimmable guard/category is needed because NativeAOT uses the trimmable typemap by default.

**Review by dalexsoto — APPROVED** [review]

File: src/Mono.Android/Java.Interop/SafeJavaCollectionFactory.cs [hunk 6]
```diff
@@ -61,170 +107,139 @@ static bool IsSupportedCollectionShape (CollectionShape shape)
...
+			return null;
 		}
 
-		var genericDefinition = targetType.GetGenericTypeDefinition ();
-		if (genericDefinition == typeof (IList<>) || genericDefinition == typeof (JavaList<>)) {
-			shape = new CollectionShape (JavaCollectionKind.List, targetType.GetGenericArguments ());
-			return true;
+		if (elementType.IsValueType) {
+			var valueFactory = ValueTypeFactory.PrimitiveTypeFactories [elementType];
+			return valueFactory.CreateList (handle, transfer);
```
[the remark this answers is not in the document]
Comment (jonathanpeppers) [reply src/Mono.Android/Java.Interop/SafeJavaCollectionFactory.cs]: Fixed in `ac59af52a`. Value factories and dictionary argument shapes are now resolved once in `TryGetFromJniHandleConverter`; the resulting delegates retain only the null-handle check and direct construction call.

File: tests/Mono.Android-Tests/Mono.Android-Tests/Java.Interop/JavaConvertTest.cs [hunk 7]
```diff
@@ -221,6 +221,58 @@ public void FromJniHandle_IListByte ()
...
+		public void FromJniHandle_ReferenceArgumentsUseCanonicalTemplate (Type targetType, Type expectedWrapperDefinition, bool dictionary)
+		{
+			Java.Lang.Object source = dictionary ? new JavaDictionary () : new JavaList ();
+			using (source) {
+				var converted = InvokeJavaConvertFromJniHandle (targetType, source.Handle, JniHandleOwnership.DoNotTransfer);
+				try {
+					var convertedType = converted.GetType ();
+					Assert.AreEqual (expectedWrapperDefinition, convertedType.GetGenericTypeDefinition ());
+					CollectionAssert.AreEqual (targetType.GetGenericArguments (), convertedType.GetGenericArguments ());
+					Assert.IsTrue (targetType.IsInstanceOfType (converted));
```
[the remark this answers is not in the document]
Comment (jonathanpeppers) [reply tests/Mono.Android-Tests/Mono.Android-Tests/Java.Interop/JavaConvertTest.cs:238]: Thanks. I am keeping this test focused on canonical-template activation: successfully constructing the exact closed wrapper type already proves that the activation constructor was preserved. Element access exercises the separate element-marshaling path, and using the closed generic wrapper directly in this test could also root the instantiation that this reflection-only regression is intended to validate.

**Review by rolfbjarne — APPROVED** [review]
